convert.py

```diff
@@ -15,65 +15,114 @@
 import os
 import argparse
 
-from onnx import helper
+from onnx import helper, checker
 import paddle.fluid as fluid
 
-import ops
-from variables import paddle_variable_to_onnx_tensor
+import fluid_onnx.ops as ops
+from fluid_onnx.variables import paddle_variable_to_onnx_tensor
+from fluid_onnx.variables import PADDLE_TO_ONNX_DTYPE
 
 
-def convert(dirname):
+def parse_args():
+    # Read arguments: path to model.
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--fluid_model", required=True, help="Input PaddlePaddle Fluid model.")
+    parser.add_argument(
+        "--onnx_model_dir",
+        required=False,
+        help="The output directory for ONNX model.")
+    args = parser.parse_args()
+    return args
+
+
+def convert(args):
     # Read the model files.
     place = fluid.CPUPlace()
     exe = fluid.Executor(place)
 
     inference_scope = fluid.core.Scope()
     with fluid.scope_guard(inference_scope):
         [inference_program, feed_target_names,
-         fetch_targets] = fluid.io.load_inference_model(dirname, exe)
+         fetch_targets] = fluid.io.load_inference_model(args.fluid_model, exe)
 
         # Using blocks in programs, create nodes using:
         onnx_nodes = []
-        all_inputs = []
-        for block in inference_program.blocks:
-            all_inputs += [
-                paddle_variable_to_onnx_tensor(v, block) for v in block.vars
-                if v not in ['feed', 'fetch']
-            ]
 
+        # Load parameters
+        global_block = inference_program.global_block()
+        for var_name in global_block.vars:
+            var = global_block.var(var_name)
+            if var_name not in ['feed', 'fetch'] and var.persistable:
+                param = fluid.executor.fetch_var(var_name, inference_scope)
```

Comment:
Oh man this is great, I spent so much time writing a custom (py)binding to deserialize params manually, I'll throw that stuff for now. `fetch_var` is beautiful. You are the expert

Reply:
No. Actually I am also not familiar with this part, and just find this method by chance :-)

Reply:
So let me tell why I have been trying to find overly complex solutions to this. If you see [this](https://github.com/PaddlePaddle/Paddle/blob/6cfc0c14971828ee9528502a2787456869210a5c/paddle/fluid/framework/executor.cc#L336), it clearly seems like scopes get destroyed after their runs - which is fair, we shouldnt have these thousands, sometimes millions, of variables in memory if they are unused. The only things saved from destruction are persistable global block things (which you seem to be using here). So vars not in global blocks can't be fetched because everything other than global block vars are destroyed.

Now on second thoughts, we can go through the global block only because we don't care about inner blocks, given our use case. But we should be careful because `fetch_var` would not have been an option if we had cared about them.

It's also important to be careful because the suggested way to fetch anything out of a program run is to add it to the `fetch_list` argument - and not `fetch_var`. But to be able to use that, we can't use the default `load_inference_model` model function - we'd need to tweak it to pass in and return `fetch_list` we are interested in.

Reply:
Oh you have a much deeper survey than me.